oak/server/oak_runtime.cc

```diff
@@ -133,12 +169,20 @@ int32_t OakRuntime::GetPort() { return grpc_node_->GetPort(); }
 
 grpc::Status OakRuntime::Stop() {
   LOG(INFO) << "Stopping runtime...";
-  grpc_node_ = nullptr;
-  for (auto& named_node : nodes_) {
+
+  // Take local ownership of all the nodes so we can stop them without holding
+  // the runtime lock.
+  std::map<std::string, std::unique_ptr<OakNode>> nodes;
+  {
+    absl::MutexLock lock(&mu_);
+    grpc_node_ = nullptr;
+    nodes = std::move(nodes_);
```

Comment:
Quite late to the party, sorry :)
Why do you move these here? Can we just call Stop on all of them?

Reply:
We do call `Stop()` on all of them just below, but we need to do it without holding `OakRuntime.mu_` &ndash; there's a chance that a per-Node thread might be just about to do a `node_create()` operation, which would in turn require the runtime lock (and so would deadlock if we did the naive thing here).

Added an extra comment.